c10d/logging: add C10D_LOCK_GUARD

d4l3k opened the pull request:
This adds logs if we can't acquire locks in NCCLUtils and ProcessGroupNCCL for 30s.

This is motivated by some deadlocks were seeing and it's unclear if it's in NCCL or on the PyTorch side of things.

This required replacing most `std::mutex` with `std::timed_mutex` and `std::condition_variable_any` as appropriate.

Test plan:

existing CI for regressions

will add unit tests on `C10D_LOCK_GUARD`


cc @XilunWu @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @c-p-i-o @gujinghui @PenghuiCheng @XiaobingSuper @jianyuh @jgong5 @mingfeima @sanchitintel @ashokei @jingxu10 @min-jean-cho @yanbing-j @Guobing-Chen @Xia-Weiwen @snadampal

fduwjj commented on torch/csrc/distributed/c10d/logging.cpp:
```diff
@@ -34,4 +34,20 @@ bool isLogLevelEnabled(LogLevel level) noexcept {
+    c10::string_view desc,
+    c10::string_view file,
+    int line) {
+  while (!lock.try_lock_for(log_interval)) {
```
I am wondering if we can just do `try_lock_for` just once and then try to capture the lock like what we do in the past?

d4l3k replied:
yeah, that's fair -- just depends if it's too much log spam

c-p-i-o commented on torch/csrc/distributed/c10d/Work.cpp:
```diff
@@ -45,17 +46,17 @@ OpType Work::retrieveOpType() const {
 
 bool Work::isSuccess() const {
-  std::lock_guard<std::mutex> lock(mutex_);
+  C10D_LOCK_GUARD(lock, mutex_);
```
wonder if it makes sense to change the name `lock` to instead be more descriptive?
We can do that later too.

d4l3k replied:
Right now we don't print the lock name -- just the mutex

This variable name only is used when condition variables are used as well

d4l3k commented:
@pytorchbot merge

d4l3k commented:
@pytorchbot merge

fduwjj reviewed (approved):
Test looks great!

d4l3k commented:
@pytorchbot merge

ZainRizvi commented:
@pytorchbot revert -c ghfirst -m "Sorry but this causes formatting errors internally which make it fail to build. See D61759282"

xwang233 commented:
Thanks @ZainRizvi  for reverting the PR. Our build also failed with the header issue. cc @malfet @atalman 

```
#45 84.81 In file included from /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/distributed/c10d/NCCLUtils.hpp:17,
#45 84.81                  from /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/distributed/c10d/ProcessGroupNCCL.hpp:22,
#45 84.81                  from xxx:
#45 84.81 /usr/local/lib/python3.10/dist-packages/torch/include/torch/csrc/distributed/c10d/logging.h:14:10: fatal error: fmt/format.h: No such file or directory
#45 84.81    14 | #include <fmt/format.h>
#45 84.81       |          ^~~~~~~~~~~~~~
#45 84.81 compilation terminated.
```

d4l3k commented:
@xwang233 you likely need to add fmt as a dependency to your custom build -- we depend on fmt elsewhere in torch but not in torch_cuda/rocm until now.

Will reland this once CI passes

d4l3k commented:
Ah -- we're leaking fmt into the public headers, will fix

d4l3k committed:
c10d/logging: add C10D_LOCK_GUARD (#134131)

Summary:
This adds logs if we can't acquire locks in NCCLUtils and ProcessGroupNCCL for 30s.

This is motivated by some deadlocks were seeing and it's unclear if it's in NCCL or on the PyTorch side of things.

This required replacing most `std::mutex` with `std::timed_mutex` and `std::condition_variable_any` as appropriate.

Test Plan:
existing CI for regressions

will add unit tests on `C10D_LOCK_GUARD`

cc XilunWu H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab c-p-i-o gujinghui PenghuiCheng XiaobingSuper jianyuh jgong5 mingfeima sanchitintel ashokei jingxu10 min-jean-cho yanbing-j Guobing-Chen Xia-Weiwen snadampal

Reviewed By: c-p-i-o

Differential Revision: D61803744

Pulled By: d4l3k

kit1980 commented:
This was reverted by D61985186

kit1980 commented:
@pytorchbot revert -m "See D61985186" -c ghfirst
